src/main/scala/strawman/collection/immutable/Range.scala

```diff
@@ -0,0 +1,280 @@
+  *  `init`) are also permitted on overfull ranges.
+  *
+  *  @param start       the start of this range.
+  *  @param end         the end of the range.  For exclusive ranges, e.g.
+  *                     `Range(0,3)` or `(0 until 3)`, this is one
+  *                     step past the last one in the range.  For inclusive
+  *                     ranges, e.g. `Range.inclusive(0,3)` or `(0 to 3)`,
+  *                     it may be in the range if it is not skipped by the step size.
+  *                     To find the last element inside a non-empty range,
+  *                     use `last` instead.
+  *  @param step        the step for the range.
+  *  @param isInclusive whether the end of the range is included or not
+  */
+@SerialVersionUID(7618862778670199309L)
+final class Range(
+  val start: Int,
+  val end: Int,
+  val step: Int,
+  val isInclusive: Boolean
+)
+  extends IndexedSeq[Int]
+    with IterableFactory[IndexedSeq]
+    with IterablePolyTransformsFromIterable[Int, IndexedSeq]
+    with SeqMonoTransformsFromIterable[Int, IndexedSeq[Int]]
+    with Serializable { range =>
+
+  def iterator(): Iterator[Int] =
+    new Iterator[Int] {
+      private var current: Seq[Int] = range
+      def next() = {
+        val x = current.head
+        current = current.tail
+        x
+      }
+      def hasNext = !current.isEmpty
+    }
+
+  private def gap           = end.toLong - start.toLong
+  private def isExact       = gap % step == 0
+  private def hasStub       = isInclusive || !isExact
+  private def longLength    = gap / step + ( if (hasStub) 1 else 0 )
+
+  override val isEmpty: Boolean = (
+    (start > end && step > 0)
+      || (start < end && step < 0)
+      || (start == end && !isInclusive)
+    )
+
+  private val numRangeElements: Int = {
+    if (step == 0) throw new IllegalArgumentException("step cannot be 0.")
+    else if (isEmpty) 0
+    else {
+      val len = longLength
+      if (len > scala.Int.MaxValue) -1
+      else len.toInt
+    }
+  }
+
+  // This field has a sensible value only for non-empty ranges
+  private val lastElement = step match {
+    case 1  => if (isInclusive) end else end-1
+    case -1 => if (isInclusive) end else end+1
+    case _  =>
+      val remainder = (gap % step).toInt
+      if (remainder != 0) end - remainder
+      else if (isInclusive) end
+      else end - step
+  }
+
+  /** The last element of this range.  This method will return the correct value
+    *  even if there are too many elements to iterate over.
+    */
+  def last: Int = if (isEmpty) Nil.head else lastElement
+  override def head: Int = if (isEmpty) Nil.head else start
+
+  /** Creates a new range containing all the elements of this range except the first one.
+    *
+    *  $doesNotUseBuilders
+    *
+    *  @return  a new range consisting of all the elements of this range except the first one.
+    */
+  override def tail: Range = {
+    if (isEmpty)
+      Nil.tail
+
+    drop(1)
```

Comment:
Going through `drop(1)` does a huge amount of unnecessary computation.  We already know it's not empty.  We need only `if (numRangeElements == 1) newEmptyRange(end) else new Range(start + step, end, step, isInclusive)`